libs/template/renderer.go

```diff
@@ -0,0 +1,260 @@
+	// outcomes for skip:
+	//
+	// 1. File is not generated. This happens if one of the file's parent directories
+	// match a glob pattern
+	//
+	// 2. File is generated but not persisted to disk. This happens if the file itself
+	// matches a glob pattern, but none of it's parents match a glob pattern from the list
+	//
+	// 3. File is persisted to disk. This happens if the file and it's parent directories
+	// do not match any glob patterns from this list
+	skipPatterns []string
+
+	// Filer rooted at template root. The file tree from this root is walked to
+	// generate the project
+	templateFiler filer.Filer
+
+	// Filer rooted at the project instance root. Allows writing files generated
+	// when walking the template file tree
+	instanceFiler filer.Filer
+}
+
+func newRenderer(ctx context.Context, config map[string]any, libraryRoot, instanceRoot, templateRoot string) (*renderer, error) {
+	// Initialize new template, with helper functions loaded
+	tmpl := template.New("").Funcs(HelperFuncs)
+
+	// Load user defined associated templates from the library root
+	libraryGlob := filepath.Join(libraryRoot, "*")
+	matches, err := filepath.Glob(libraryGlob)
+	if err != nil {
+		return nil, err
+	}
+	if len(matches) != 0 {
+		tmpl, err = tmpl.ParseGlob(libraryGlob)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	templateFiler, err := filer.NewLocalClient(templateRoot)
+	if err != nil {
+		return nil, err
+	}
+
+	instanceFiler, err := filer.NewLocalClient(instanceRoot)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx = log.NewContext(ctx, log.GetLogger(ctx).With("action", "initialize-template"))
+
+	return &renderer{
+		ctx:           ctx,
+		config:        config,
+		baseTemplate:  tmpl,
+		files:         make([]*inMemoryFile, 0),
+		skipPatterns:  make([]string, 0),
+		templateFiler: templateFiler,
+		instanceFiler: instanceFiler,
+	}, nil
+}
+
+// Executes the template by applying config on it. Returns the materialized template
+// as a string
+func (r *renderer) executeTemplate(templateDefinition string) (string, error) {
+	// Create copy of base template so as to not overwrite it
+	tmpl, err := r.baseTemplate.Clone()
+	if err != nil {
+		return "", err
+	}
+
+	// Parse the template text
+	tmpl, err = tmpl.Parse(templateDefinition)
+	if err != nil {
+		return "", err
+	}
+
+	// Execute template and get result
+	result := strings.Builder{}
+	err = tmpl.Execute(&result, r.config)
+	if err != nil {
+		return "", err
+	}
+	return result.String(), nil
+}
+
+func (r *renderer) computeFile(relPathTemplate string) (*inMemoryFile, error) {
+	// read template file contents
+	templateReader, err := r.templateFiler.Read(r.ctx, relPathTemplate)
+	if err != nil {
+		return nil, err
+	}
+	contentTemplate, err := io.ReadAll(templateReader)
+	if err != nil {
+		return nil, err
+	}
+
+	// execute the contents of the file as a template
+	content, err := r.executeTemplate(string(contentTemplate))
```

Comment:
This means we also pass binaries/non-template files into the template execution.

I'm thinking we may need to revisit a `.tmpl` extension for templates because files with binary data likely won't parse as a template, or worse, parse as a template and produce some invalid output.

Reply:
Doing this as a followup